### ISSUE TITLE
Extension of `iterate`: any arity

### ISSUE THREAD

**Issue by nanouasyn** [more-itertools/more-itertools#721]
Let the `iterate` function use `func` of any arity, not only unary `func`.

If a unary `func`, using the initial element `x`, generates a sequence (`x`, `func(x)`, `func(func(x))`, ...), then a binary `func`, using two initial elements `x` and `y`, should generate a sequence (`x1`, `x2`,` func(x1, x2)`, `func(x2, func(x1, x2))`, `func(func(x1, x2)`, `func(x2, func(x1, x2)))`, ...). If we generalize it, an `N`-ary function, using the initial elements `x1`, `x2`, ..., `xN`, should generate a sequence (`x1`, ..., `xN`, `func(x1, ..., xN)`, `func(x2, ..., xN, func(x1, ..., xN))`, `func(x3, ..., xN, func(x1, ..., xN), func(x2, ..., xN, func(x1, ..., xN)))` ...).

So, `iterate` function should accept a variable number of initial elements. Their number will determine how many previous elements of the sequence the `func` function should receive in order to generate the next element of the sequence. To apply an `N`-ary `func`, you only need to pass this function and the first `N` elements of sequence. So, `iterate(func, x)` will use unary `func`, `iterate(func, x1, x2)` will use binary `func`, and so on. Similarly, the number of arguments affects the arity in the `map` function.

If we can use binary functions, we can, for example, express the Fibonacci numbers in one line:

```python
>>> take(10, iterate(lambda a, b: a + b, 0, 1))
[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
```

Or triangular numbers, if we use a ternary function and three initial elements:

```python
>>> take(10, iterate(lambda a, b, c: 3*c-3*b+a, 0, 1, 3))
[0, 1, 3, 6, 10, 15, 21, 28, 36, 45]
```

A large number of useful sequences are defined in a similar recursive way, when the next element of the sequence is determined by some fixed number of previous ones. Therefore, such a change will significantly expand the use cases of `iterate`.

My suggestion is something like this:

```python
def iterate(func, *initials):
	yield from initials
    queue = deque(initials, maxlen=len(initials))
    while True:
        queue.append(func(*queue))
        yield queue[-1]
```

I think we should also handle 0-ary functions for consistency. `iterate(func)` will be an equivalent of `repeatfunc(func)`.

Of course, we need also to take care of the feature mentioned in #707. Therefore, all `func` calls must be wrapped in a `try`/`except` block.

In addition, there is a slight backward incompatibility, because previously users could call `iterate` with the initial value passed by the `start` keyword.

I suggest using the signature `def iterate(func, *initials, start=_marker)`. If the user tries to pass positional arguments and `start` argument at the same time, we will raise `TypeError`. In future versions, it will be possible to mark the `start` parameter as deprecated, and then remove it from signature.

An alternative solution would be a signature `def iterate(func, start=_marker, *initials)` or even signature `def iterate(func, start, *initials)` (prohibit 0-ary functions). However, this will make documentation much more  complicated and confusing. After all, `start` and `initials` are conceptually one parameter, randomly divided into two parameters. It's easier to explain that there are two ways to call `iterate`: either with positional arguments or with keyword. In addition, such a signature will prevent deleting the parameter in future versions if we still decide to do it.

Another alternative would be to create a new function, for example, `unreduce`. However, for a number of reasons, this solution does not seems to be successful enough. Yet another function! Which most often does the same thing as `iterate`! With the same idea as `iterate` and almost the same name! I mean, there is no reason, why one of the functions is called `iterate` and the other `unreduce`, and not vice versa. This is different from the relationship between, for example, `sum` and `reduce` or `enumerate` and `zip`.

So, the final draft:
```python
def iterate(func, *initials, start=_marker):
    if start is not _marker:
        if initials:
            raise TypeError('cannot mix *start* and *initials* args')
        initials = [start]
    if not initials:
        while True:
            try:
                yield func()
            except StopIteration:
                return
    yield from initials
    queue = deque(initials, maxlen=len(initials))
    while True:
        try:
            next_element = func(*queue)
        except StopIteration:
            return
        yield next_element
        queue.append(next_element)
```

**Comment by pochmann** [comment]
I think I also did this recently, also with Fibonacci as motivating case. The triangular numbers case looks bloated, `iterate(lambda a, b: 2*b-a+1, 0, 1)` suffices. Or `accumulate(count())` :-)

**Comment by nanouasyn** [comment]
> I think I also did this recently, also with Fibonacci as motivating case. The triangular numbers case looks bloated, `iterate(lambda a, b: 2*b-a+1, 0, 1)` suffices. Or `accumulate(count())` :-)

You're right! I just took this example from [wikipedia](https://en.wikipedia.org/wiki/Constant-recursive_sequence), where it was the only example using the three previous elements to calculate the next one :) If my PR is accepted, could you find a better example of such a sequence, and replace my example in the docstring in a separate PR?

**Comment by pochmann** [comment]
Found mine again, it's different. Instead of a binary function, it wants a unary function that accepts sequence pairs. So Fibonacci is `extend((0, 1), sum)`. Binary / n-ary function is probably nicer in general, easier to write lambdas.

**Comment by pochmann** [comment]
Did you benchmark yours? I imagine it makes the current unary use case quite a bit slower.

**Comment by bbayles** [comment]
Thanks for the suggestion! I'm disinclined to accept this change, because I feel like you're going to be writing your a Lambda function for `iterate` anyway. Why not have it operate on tuples?

Fibonacci in one line with `iterate` as is:

```python
In [11]: take(10, (f[1] for f in iterate(lambda x: (x[1], x[0] + x[1]), (1, 1))))
Out[11]: [1, 2, 3, 5, 8, 13, 21, 34, 55, 89]
```

**Comment by nanouasyn** [comment]
I've been thinking about it too. But when we read the solution with tuples, we see that something complicated is going on. A sequence of tuples is generated from the initial tuple. What is the principle of calculating the next tuple? The second element of the tuple becomes the first, and the second becomes the sum of the elements of the tuple. Finally, we extract the second element of each tuple. This tuple does not have a meaningful name and we operate with indexes of these incomprehensible tuples. We need to think carefully before we understand what the author wanted to write. As a result, there are two mistakes in your solution. It happened precisely because the code is too complex. How long it will take to find them among all these indexes and parentheses?

Also, with this approach, we calculated an extra element that no one asked us to calculate. Because we have to take the first elements of each tuple, but the second one is also calculated:

```python
>>> take(10, (x[0] for x in iterate(lambda x: (x[1], x[0] + x[1]), (0, 1))))
[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
```

Or we should further complicate the example to avoid this:

```python
>>> take(10, chain([0], (x[1] for x in iterate(lambda x: (x[1], x[0] + x[1]), (0, 1)))))
[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
```

The sequence definition alone barely fits into the PEP8 limits on the line length. We can assign this sequence to a variable, but no more than two characters will remain for its name. This is not enough even for "fib". We definitely need two lines. It's not cool.

When we read the code below, we immediately realize that a sequence starting with `0` and `1` is being created by applying the addition operation. It reads even easier than the recursive definition of Fibonacci numbers. It's as simple as `reduce`. We just write: give us the sequence `0, 1, ...` and then use addition.

```python
>>> from operator import add
>>> take(10, iterate(add, 0, 1))
[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
```

Even if users abuse lambdas, these lambdas will be simpler and clearer. The solution with upgraded `iterate` is still twice as short and twice as less nested, even if we used lambda:

```python
>>> take(10, iterate(lambda a, b: a + b, 0, 1))
[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
```

Imagine a `map` without the ability to pass an arbitrary number of sequences and without a `starmap`. It is quite possible to use `zip` and write a lambda function that will operate with tuples and indexes. But it's inconvenient. Compare this

```python
map(lambda x: x[0] + x[1], zip(it1, it2))
```

with this

```python
map(add, it1, it2)
```

Again, compare this

```python
chain([0], (x[1] for x in iterate(lambda x: (x[1], x[0] + x[1]), (0, 1)))
```

with this

```python
iterate(add, 0, 1)
```

So, for `iterate`, it means even more. If only one could be selected, I would add varargs to `iterate`, not to `map`. And in addition, as far as I understand, binary functions are used here as often as unary ones, unlike `map`. That is, you have to write a uglier solution more often.

Many functions in `more_itertools` can be replaced with a one-line expressions, but this is not a reason to abandon them. Much more important is how much more expressive the code turns out to be if you use these functions. With this change, `iterate` becomes much more expressive. Some sequences can be expressed in terms of `iterate`, even if we don't have to do it. But it gives a different way of thinking, more natural. For example,

```python
# Arithmetic and geometric progressions are almost the same.
>>> from functools import partial
>>> arithmetic_seq = partial(iterate, lambda a, b: 2 * b - a)
>>> geometric_seq = partial(iterate, lambda a, b: b ** 2 / a)

# Just 1, 2, ...
>>> take(10, arithmetic_seq(1, 2))
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# We don't actually need a numeric_range function to use floats.
>>> take(10, arithmetic_seq(1.0, 1.5))
[1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5.0, 5.5]

# Something doubles every time.
>>> take(10, geometric_seq(1.0, 2.0))
[1.0, 2.0, 4.0, 8.0, 16.0, 32.0, 64.0, 128.0, 256.0, 512.0]

# Something is reduced by half every time.
>>> take(10, (round(x, 3) for x in geometric_seq(1.0, 0.5)))
[1.0, 0.5, 0.25, 0.125, 0.062, 0.031, 0.016, 0.008, 0.004, 0.002]

# Factorials are like a geometric progression.
>>> take(10, iterate(lambda a, b: b ** 2 // a + b, 1, 1))
[1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880]

# Triangular numbers are like factorials, but for arithmetic progression.
>>> take(10, iterate(lambda a, b: 2 * b - a + 1, 0, 1))
[0, 1, 3, 6, 10, 15, 21, 28, 36, 45]
```

And we pay almost nothing for it. Just one keyword-only parameter explained in the documentation. If you want to avoid backward incompatibility at all costs, we can stop at the proposed signature and never mark a `start` parameter as deprecated. Let it be a special way to call the function. If you don't like the performance of generalized `iterate`, we can handle unary functions in separate `if` block, so it will be as fast as it was. Perhaps you have some other objections that I haven't thought of?

**Comment by bbayles** [comment]
Thanks for the writeup!

I've been loosely consideration additions to `more-itertools` on the basis of whether they're at least one of:
* Practical
* Educational
* Cool

I think the current version of `iterate` is at least practical (it's practical in Haskell, anyway; that's [where it came from](https://github.com/more-itertools/more-itertools/commit/99c3ed0aae283afe8964087d24eb13febd5ed6b9)). But I'm not sure if giving it more features is particularly practical or educational. Maybe it's cool?

If we can find some example of something beyond toy math problems, I would be more inclined toward this sort of extension.

**Comment by pochmann** [comment]
> But it gives a different way of thinking, more natural. For example,

Hmm, all of those examples look *less* natural than already available alternatives. ([Attempt This Online!](https://ato.pxeger.com/run?1=bVFBTgMxDBTXvMK3TVBYdZGQUKX-gg-ku17VIpusEqe0b-HSC3yBv_AaskmXcsA32zNjj_3-OZ_54N3l8pF4fHj-vvsag5-AGAN7byPQNPvAYPo-TckaRg29T441ULTUoyh4P2Mw7MMKz1AhBhyBzStKp4ug2VtUWwE5AnIKDixFllVHrggNTqlKLiVGOSbXa4hsAl_5bweyCC8hYc2XOBPaoaJ-ayWDHSwKsgoIscjGXGyaRhQvsms3GjbtkxLrSGum_WDgtIUT3MOjhgxR4nYFeSUqvXjN_hwxGbvr_gFlO3lUGznQLFUbZ0tsyWGUeZlxORpnfnEbq505UOZRXvaWlUt2eU88Grv01N-uqv-7vnF95w8))

```python
count(1.0, 0.5)
[1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5.0, 5.5]

iterate(lambda x: x * 2, 1.0)
[1.0, 2.0, 4.0, 8.0, 16.0, 32.0, 64.0, 128.0, 256.0, 512.0]

accumulate(count(1), mul, initial=1)
[1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880]

accumulate(count())
[0, 1, 3, 6, 10, 15, 21, 28, 36, 45]
```

Your arithmetic/geometric sequence might also be less accurate, as you *compute* the difference/multiplier, from computed values (which can be inaccurate). And that extra computation also costs extra time.

**Comment by nanouasyn** [comment]
> Hmm, all of those examples look _less_ natural than already available alternatives.

It's more natural in another sense :) Perhaps I expressed myself incorrectly. I meant that many sequences, which are usually obtained in different ways, can be uniformly expressed using `iterate`. That is, these sequences seem fundamentally different, but when we write them using `iterate`, we see the general principle. Just by changing the formula a little, we get very different things. So, specifying the formula and initial elements seems like a more natural way to think about these sequences. Indeed, your alternatives are more understandable, obvious and preferable. It's like using `sum`, `math.prod` or `all` instead of `reduce`. They better express our intentions, so they should be preferred. But `reduce` expresses a more general principle, and we can always use it if we don't have the specialized function we need.

> Maybe it's cool?

Yes, I think so :) The ability to express Fibonacci numbers as `iterate(add, 0, 1)` looks cool for me. And it seems to be a natural generalization of `iterate`. However, I ran out of arguments in favor of the coolness of this decision. If I still haven't managed to convince you, we probably have different views on coolness.

> If we can find some example of something beyond toy math problems, I would be more inclined toward this sort of extension.

I doubt that we will be able to find something beyond toy math problems. I've never seen anyone use Fibonacci numbers or anything like that to solve real problems. It seems to me that Fibonacci numbers are needed only to learn something about Fibonacci numbers. For example, I noticed that there seem to be too many odd numbers. I checked how many such numbers are in the first hundred numbers, in the first thousand, in the first ten thousand. And it turned out that odd numbers seem to be twice as many as even ones. It can't be a coincidence. I started thinking, and found that this sequence is even, odd, odd, even, odd, odd, ... Indeed, even + odd = odd, odd + odd = even, odd + even = odd. This is an interesting fact. I had fun, but it had no practical value. On the other hand, I had a specific problem, and being able to generate Fibonacci numbers and explore them proved to be a useful tool to solve it. Perhaps people who are engaged in programming closer to science will be able to give examples of more practically useful problems. Сan we leave this issue open to wait for such people?
